Src/IronPython.Modules/_codecs.cs

```diff
@@ -172,67 +172,28 @@ public static PythonTuple ascii_encode(CodeContext context, string input, string
         #region Escape Encoding
 
         public static PythonTuple escape_decode(CodeContext/*!*/ context, string data, string errors = "strict")
-            => escape_decode(DoEncode(context, "utf-8", Encoding.UTF8, data, "strict").Item1, errors);
+            => escape_decode(StringOps.DoEncodeUtf8(context, data), errors);
 
         public static PythonTuple escape_decode([BytesConversion]IList<byte> data, string errors = "strict") {
-            var res = new StringBuilder();
-            for (int i = 0; i < data.Count; i++) {
-                if (data[i] == '\\') {
-                    if (i == data.Count - 1) throw PythonOps.ValueError("\\ at end of string");
-
-                    switch ((char)data[++i]) {
-                        case 'a': res.Append((char)0x07); break;
-                        case 'b': res.Append((char)0x08); break;
-                        case 't': res.Append('\t'); break;
-                        case 'n': res.Append('\n'); break;
-                        case 'r': res.Append('\r'); break;
-                        case '\\': res.Append('\\'); break;
-                        case 'f': res.Append((char)0x0c); break;
-                        case 'v': res.Append((char)0x0b); break;
-                        case '\n': break;
-                        case 'x':
-                            if (++i < data.Count && CharToInt((char)data[i], out int dig1)
-                                    && ++i < data.Count && CharToInt((char)data[i], out int dig2)) {
-                                res.Append((char)(dig1 * 16 + dig2));
-                            } else {
-                                switch (errors) {
-                                    case "strict":
-                                        throw PythonOps.ValueError("invalid \\x escape at position {0}", i);
-                                    case "replace":
-                                        res.Append("?");
-                                        i--;
-                                        break;
-                                    default:
-                                        throw PythonOps.ValueError("decoding error; unknown error handling code: " + errors);
-                                }
-                            }
-                            break;
-                        default:
-                            res.Append("\\" + (char)data[i]);
-                            break;
-                    }
-                } else {
-                    res.Append((char)data[i]);
+            var res = LiteralParser.ParseBytes(data, 0, data.Count, isRaw: false, normalizeLineEndings: false, errorHandler);
+
+            return PythonTuple.MakeTuple(Bytes.Make(res.ToArray()), data.Count);
+
+            IReadOnlyList<byte> errorHandler(IList<byte> data, int start, int end) {
```

Comment:
Not sure how I feel about having to do a string switch statement to check `errors` every time this is called. I guess we can leave it for now since you'll probably end up re-working the pattern if you want to do something similar in `ParseString`...

Reply:
Thank you for the comment. Yes, `ParseString` with `unicode-decode` will be more demanding and I will probably need to modify the pattern. But since this case is simpler, I'd rather have a starting solution here that we both feel good about. In the next commit I will propose a variant that does the string switch at most only once.

Reply:
Maybe something like this would make it slightly easier to follow the logic?
```C#
LiteralParser.ParseBytesErrorHandler<byte> getErrorHandler(string errors) {
    if (errors == null) return null;
    Func<int, IReadOnlyList<byte>> eh = null;
    return errorHandler;

    IReadOnlyList<byte> errorHandler(IList<byte> data, int start, int end) {
        eh ??= errors switch
        {
            "strict" => idx => throw PythonOps.ValueError("invalid \\x escape at position {0}", idx),
            "replace" => idx => _replacementMarker ??= new[] { (byte)'?' },
            "ignore" => idx => null,
            _ => idx => throw PythonOps.ValueError("decoding error; unknown error handling code: " + errors),
        };
        return eh(start);
    }
}
```

Reply:
Good idea. Especially that for `unicode-escape`, `getErrorHandler` will only get more complex.